src/auth/impersonated.ts

```diff
@@ -0,0 +1,213 @@
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {GaxiosError, GaxiosOptions, GaxiosPromise} from 'gaxios';
+import {
+  GetTokenResponse,
+  Headers,
+  OAuth2Client,
+  RefreshOptions,
+  RequestMetadataResponse,
+} from './oauth2client';
+
+export interface ImpersonatedOptions extends RefreshOptions {
+  sourceClient?: OAuth2Client;
+  targetPrincipal?: string;
+  targetScopes?: string[];
+  delegates?: string[];
+  lifetime?: number | 3600;
+}
+
+export interface TokenResponse {
+  accessToken: string;
+  expireTime: string;
+}
+
+export class Impersonated extends OAuth2Client {
+  private sourceClient: OAuth2Client;
+  private targetPrincipal: string;
+  private targetScopes: string[];
+  private delegates: string[];
+  private lifetime: number;
+
+  /**
+   * Impersonated service account credentials.
+   *
+   * Create a new access token by impersonating another service account.
+   *
+   * Impersonated Credentials allowing credentials issued to a user or
+   * service account to impersonate another. The source project using
+   * Impersonated Credentials must enable the "IAMCredentials" API.
+   * Also, the target service account must grant the orginating principal
+   * the "Service Account Token Creator" IAM role.
+   *
+   * @param credentials the service account email address.
+   * @param sourceClient the source credential used as to acquire the
+   * impersonated credentials
+   * @param targetPrincipal the service account to impersonate.
+   * @param delegates the chained list of delegates required to grant the
+   *  final access_token. If set, the sequence of identities must have
+   * "Service Account Token Creator" capability granted to the preceding
+   * identity. For example, if set to [serviceAccountB, serviceAccountC],
+   * the sourceCredential must have the Token Creator role on serviceAccountB.
+   * serviceAccountB must have the Token Creator on serviceAccountC.
+   * Finally, C must have Token Creator on target_principal.
+   * If left unset, sourceCredential must have that role on targetPrincipal.
+   * @param targetScopes scopes to request during the authorization grant.
+   * @param lifetime number of seconds the delegated credential should be
+   * valid for (up to 3600).
```

Comment:
If I need to hold the actual credentials to generate the impersonated client, how does this provide improved security. I think I just don't fully understand the threat model this protects me from (not be a security person).

Reply:
You'll always need some credentials to bootstrap with but this capability basically allows you to impersonate _another_ credential

you can do something like this:  You own serivceAccoutnA, I own  serviceAccountB and I give B permissions on GCS BucketG that i own; A does not have permission on G.

I can create an [IAM condition](https://cloud.google.com/iam/docs/conditions-overview) that says "only allow A access to impersonate B between 1am and 2am".  So what serviceAccountA has is temp access to bucketG.  
You can basically continue this chain of delegation.  The full flow is described here

- https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials